pkg/apis/batch/validation/validation.go

```diff
@@ -486,9 +499,115 @@ func validatePodTemplateUpdate(spec, oldSpec batch.JobSpec, fldPath *field.Path,
+	statusFld := field.NewPath("status")
+	allErrs = append(allErrs, validateJobStatus(&job.Status, statusFld)...)
+
+	// A strengthened validation of the Job status transitions is needed since the
+	// Job managed-by label let's the Job object be controlled by external
+	// controllers. We want to make sure the transitions done by the external
+	// controllers meet the expectations of the clients of the Job API.
+	// For example, we verify that a Job in terminal state (Failed or Complete)
+	// does not flip to a non-terminal state.
+	if utilfeature.DefaultFeatureGate.Enabled(features.JobManagedByLabel) {
+		for _, cType := range []batch.JobConditionType{batch.JobFailed, batch.JobComplete, batch.JobFailureTarget} {
+			if isConditionTrue(oldJob.Status.Conditions, cType) && !isConditionTrue(job.Status.Conditions, cType) {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("conditions"), field.OmitValueType{}, fmt.Sprintf("cannot disable the terminal %s=True condition", string(cType))))
+			}
+		}
+		if job.Status.Failed < oldJob.Status.Failed {
+			allErrs = append(allErrs, field.Invalid(statusFld.Child("failed"), job.Status.Failed, "cannot decrease the failed counter"))
+		}
+		isIndexed := ptr.Deref(job.Spec.CompletionMode, batch.NonIndexedCompletion) == batch.IndexedCompletion
+		if !isIndexed {
+			if job.Status.Succeeded < oldJob.Status.Succeeded {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("succeeded"), job.Status.Succeeded, "cannot decrease the succeeded counter"))
+			}
+		}
+		// In the checks below we fail validation for Job status fields (or conditions) only if they change their values
+		// (compared to the oldJob). This allows proceeding with status updates unrelated to the fields violating the
+		// checks, while blocking bad status updates for jobs with correct status.
+		if oldJob.Status.CompletionTime != nil {
+			if job.Status.CompletionTime == nil {
+				allErrs = append(allErrs, field.Required(statusFld.Child("completionTime"), "completionTime cannot be removed"))
+			} else {
+				allErrs = append(allErrs, apivalidation.ValidateImmutableField(job.Status.CompletionTime, oldJob.Status.CompletionTime, statusFld.Child("completionTime"))...)
+			}
+		}
+		if !ptr.Equal(oldJob.Status.CompletionTime, job.Status.CompletionTime) || isConditionTrue(oldJob.Status.Conditions, batch.JobComplete) != isConditionTrue(job.Status.Conditions, batch.JobComplete) {
+			if job.Status.CompletionTime != nil && !isConditionTrue(job.Status.Conditions, batch.JobComplete) {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("completionTime"), job.Status.CompletionTime, "cannot set completionTime when there is no Complete=True condition"))
+			}
+			if oldJob.Status.CompletionTime == nil && job.Status.CompletionTime == nil && isConditionTrue(job.Status.Conditions, batch.JobComplete) {
+				allErrs = append(allErrs, field.Required(statusFld.Child("completionTime"), "completionTime is required for Complete jobs"))
```

Comment:
is the `oldJob` check necessary here?

Reply:
Good question, it was added, because otherwise I get two errors when transitioning from completeTime!=nil to completionTime=nil. The other validation check returning the error is above. Semantically it does not matter, but I wanted to prevent double erroring on the same scenario. Having said that, I'm going to revisit.